src/ServiceControl.Monitoring/Raw/RawDataProvider.cs

```diff
@@ -1,29 +1,83 @@
 ﻿namespace ServiceControl.Monitoring.Raw
 {
+    using System;
     using System.Collections.Concurrent;
     using System.Collections.Generic;
-    using Http;
+    using System.Linq;
     using Newtonsoft.Json.Linq;
 
-    /// <summary>
-    /// The raw endpoint data provider, consuming data with <see cref="Consume"/> and providing them with raw data.
-    /// </summary>
-    public class RawDataProvider : IEndpointDataProvider, IRawDataConsumer
+    public class RawDataProvider : IRawDataConsumer
     {
-        public const string Name = "Raw";
-
-        /// <summary>
-        /// Consumes a new portion of data.
-        /// </summary>
         public void Consume(IReadOnlyDictionary<string, string> headers, JObject data)
         {
-            var name = headers.GetOriginatingEndpoint();
-            contexts.AddOrUpdate(name, data, (context, currentData) => data);
+            var endpointName = headers.GetOriginatingEndpoint();
+            var timestamp = data["Timestamp"].Value<DateTime>();
+            var criticalTimeMetrics = data["Timers"].First(elem => elem["Name"].Value<string>() == "Critical Time");
+            var criticalTime = criticalTimeMetrics["Histogram"]["Mean"].Value<float>();
+            var processingTimeMetrics = data["Timers"].First(elem => elem["Name"].Value<string>() == "Processing Time");
+            var processingTime = processingTimeMetrics["Histogram"]["Mean"].Value<float>();
+
+            // ordered collections are not threadsafe
+            var endpointLock = endpointLocks.GetOrAdd(endpointName, new object());
+
+            lock (endpointLock)
+            {
+                if (!criticalTimeValues.ContainsKey(endpointLock))
+                    criticalTimeValues[endpointLock] = new SortedList<DateTime, float>();
+                criticalTimeValues[endpointLock].Add(timestamp, criticalTime);
```

Comment:
@Scooletz yes, this isn't addressed yet. In general there needs to be further aggregation of received values for further usage. E.g. I was thinking about multiple collections which "feed into each other" (e.g. aggregate values of the first collection every 5 minutes, feed that single value in the level 2 collection and then truncate the first one). This would provide all the data points to visualize different time intervals with little memory usage (but wouldn't allow to "zoom" on past timespans).